File: DSA/SuryanshMishra_590012069/Day82/day81question3.cpp
#include <cstddef>
#include <stack>
#include <functional>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// Recursive reverse-inorder conversion: each node becomes node + sum of greater nodes.
TreeNode* convertBSTRecursive(TreeNode* root) {
    long long acc = 0;
    // helper defined as lambda capturing acc by reference
    function<void(TreeNode*)> dfs = [&](TreeNode* node) {
        if (!node) return;
        dfs(node->right);
        acc += node->val;
        node->val = (int)acc;
        dfs(node->left);
    };
    dfs(root);
    return root;
}

// Iterative reverse-inorder conversion using a stack.
TreeNode* convertBSTIterative(TreeNode* root) {
    long long acc = 0;
    stack<TreeNode*> st;
    TreeNode* cur = root;
    while (cur || !st.empty()) {
        while (cur) { st.push(cur); cur = cur->right; }
        cur = st.top(); st.pop();
        acc += cur->val;
        cur->val = (int)acc;
        cur = cur->left;
    }
    return root;
}

File: DSA/akashyadav_590014126/Day_81_Question_3.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def printNodesInRange(root, L, R):
    def inorder(node):
        if not node:
            return
        if node.val > L:
            inorder(node.left)
        if L <= node.val <= R:
            print(node.val, end=' ')
        if node.val < R:
            inorder(node.right)
    inorder(root)
